### PR TITLE
[DataFrame] Implementing API correct groupby with aggregation methods

### PR DESCRIPTION
**Pull request by devin-petersohn**
## What do these changes do?
Adds groupby and allows users to interact with the GroupBy object the same way they would in Pandas. 

- `groupby` implementation
- `DataFrameGroupBy` object
- `agg` / `aggregate` / `apply` for non-dictionary arguments
- Some performance improvements overall.

### CODE REVIEW

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -686,14 +700,160 @@ def add(self, other, axis='columns', level=None, fill_value=None):
...
+        else:
+            partitions = self._row_partitions
+
+        if axis == 1:
+            kwargs['axis'] = axis
+            kwargs['temp_columns'] = self.columns
+        else:
+            kwargs['temp_index'] = self.index
+
+        def remote_helper(df, arg, *args, **kwargs):
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: More descriptive names would be helpful here for readability.

**Comment by robertnishihara** [comment]
Their seem to be some test failures, e.g.,

```
____________ ERROR collecting ray/dataframe/test/test_dataframe.py _____________
[31m../../../.local/lib/python2.7/site-packages/pytest-3.5.0-py2.7.egg/_pytest/python.py:411: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
../../../.local/lib/python2.7/site-packages/py-1.5.3-py2.7.egg/py/_path/local.py:668: in pyimport
    __import__(modname)
../../../.local/lib/python2.7/site-packages/pytest-3.5.0-py2.7.egg/_pytest/assertion/rewrite.py:213: in load_module
    py.builtin.exec_(co, mod.__dict__)
python/ray/dataframe/test/test_dataframe.py:9: in <module>
    import ray.dataframe as rdf
../../../.local/lib/python2.7/site-packages/ray-0.4.0-py2.7-linux-x86_64.egg/ray/dataframe/__init__.py:31: in <module>
    from .dataframe import DataFrame  # noqa: 402
../../../.local/lib/python2.7/site-packages/ray-0.4.0-py2.7-linux-x86_64.egg/ray/dataframe/dataframe.py:29: in <module>
    from .groupby import DataFrameGroupBy
E     File "/home/travis/.local/lib/python2.7/site-packages/ray-0.4.0-py2.7-linux-x86_64.egg/ray/dataframe/groupby.py", line 48
E       *part),
E       ^
E   SyntaxError: invalid syntax[0m
```

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -686,14 +700,160 @@ def add(self, other, axis='columns', level=None, fill_value=None):
...
+            [_deploy_func._submit(args=(lambda df: remote_helper(df,
+                                                                 func,
+                                                                 *args,
+                                                                 **kwargs),
+                                        part), num_return_vals=4)
+             for part in partitions]
+
+        # This magic unzips the list comprehension returned from remote
+        is_series, new_parts, index, columns = \
+            [list(t) for t in zip(*remote_result)]
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Do you need each variable in list form? `zip` should allow auto-unboxing.

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -686,14 +700,160 @@ def add(self, other, axis='columns', level=None, fill_value=None):
...
+
+        # This part is because agg can allow returning a Series or a
+        # DataFrame, and we have to determine which here. Shouldn't add
+        # too much to latency in either case because the booleans can
+        # be returned immediately
+        is_series = ray.get(is_series)
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Wouldn't getting the booleans require a block on the rest of the parameters being calculated anyways?

File: python/ray/dataframe/groupby.py [hunk 4]
```diff
@@ -3,60 +3,93 @@
...
+
+        self._grouped_partitions = \
+            list(zip(*(groupby._submit(args=(by,
+                                             axis,
+                                             level,
+                                             as_index,
+                                             sort,
+                                             group_keys,
+                                             squeeze,
+                                             *part),
```
Comment (robertnishihara) [inline python/ray/dataframe/groupby.py]: looks like a python 2 issue, you may need to do 

```python
                                       args=(by,
                                             axis,
                                             level,
                                             as_index,
                                             sort,
                                             group_keys,
                                             squeeze) + part
```

or something like that.

Comment (robertnishihara) [reply python/ray/dataframe/groupby.py]: I can give this a try.

File: python/ray/dataframe/dataframe.py [hunk 5]
```diff
@@ -686,14 +700,160 @@ def add(self, other, axis='columns', level=None, fill_value=None):
...
+        if all(is_series):
+            new_series = pd.concat(ray.get(new_parts))
+            new_series.index = self.columns if axis == 0 else self.index
+            return new_series
+        # this happens when some of the partitions return Series and others
+        # return DataFrames
+        elif any(is_series):
+            raise ValueError("no results.")
+        elif axis == 0:
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: What's different between the last two cases in this if statement?

**Review by p-yang — CHANGES_REQUESTED** [review]

File: python/ray/dataframe/dataframe.py (hunk 2 above)
Comment (devin-petersohn) [reply python/ray/dataframe/dataframe.py]: We do need most of them in list form. I'll go ahead and change it.

File: python/ray/dataframe/dataframe.py [hunk 6]
```diff
@@ -686,14 +700,160 @@ def add(self, other, axis='columns', level=None, fill_value=None):
...
+        if isinstance(arg, compat.string_types):
+            return self._string_function(arg, *args, **kwargs)
+
+        # Dictionaries have complex behavior because they can be renamed here.
+        elif isinstance(arg, dict):
+            raise NotImplementedError(
+                "To contribute to Pandas on Ray, please visit "
+                "github.com/ray-project/ray.")
+        elif is_list_like(arg):
+            from .concat import concat
```
[the remark this answers is not in the document]
Comment (devin-petersohn) [reply python/ray/dataframe/dataframe.py]: Cyclical import won't allow this.

File: python/ray/dataframe/dataframe.py (hunk 2 above)
Comment (devin-petersohn) [reply python/ray/dataframe/dataframe.py]: We do actually need them all in lists because of the `ray.get()`

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**